File: cpp/vm/advm/repl.cpp
#include "repl.h"
#include <iostream>
#include <fstream>

Repl::Repl() {
    program = new ASTProgram();
    gc = new GarbageCollector();
    compiler.gc = gc;
    vm.gc = gc;
}

Repl::~Repl() {
    gc->forceFreeObjects();
    finalize_free_all_builtin_objects();
    delete gc;
    delete program;
}

void Repl::loop() {
    std::cout << "testing a simple VM\n";
    while(1) {
        std::string line;
        std::cout << ">> ";
        getline(std::cin, line);
        if (line == "exit()") break;
        if (line == "stack()") {
            vm.printStack();
            continue;
        }
        parser.load(line);
        program->reset();
        parser.buildProgramStatement(program);

        compiler.reset();
        compiler.compile(program);

        Bytecode bytecode = compiler.getBytecode();
        compiler.code.instructions = bytecode.instructions;
        std::cout << compiler.code.toString(); // asta pare ca functioneaza cum trebuie

        vm.load(bytecode);
        vm.run();

        AdObject* result = vm.last_popped_stack_elem();
        //AdObject* result = vm.last_popped_stack_element();
        if (result != NULL) {
            std::cout << result->inspect() << "\n";
        }

        gc->unmarkAllObjects();
        gc->markObjects(vm.stack, vm.sp);
        gc->markObjects(vm.constants);
        gc->sweepObjects();

    }
}

void Repl::executeFile(std::ifstream &target) {
    if (target.is_open()) {
        std::string line;
        std::string text;
        while (getline(target ,line)) {
            text += line + "\n";
        }
        parser.load(text);
        program->reset();
        parser.buildProgramStatement(program);
        std::cout << program->toString(); // uncomment this for ast elements printing

        compiler.reset();
        compiler.compile(program);

        Bytecode bytecode = compiler.getBytecode();
        compiler.code.instructions = bytecode.instructions;
        compiler.code.instructions.size = bytecode.instructions.bytes.size();
        std::cout << compiler.code.toString(); // uncomment this for bytecode printing

        vm.load(bytecode);
        vm.run();

        AdObject* result = vm.last_popped_stack_elem();
        //AdObject* result = vm.last_popped_stack_element();
        if (result != NULL) {
            std::cout << result->inspect() << "\n";
        }

        gc->unmarkAllObjects();
        gc->markObjects(vm.stack, vm.sp);
        gc->markObjects(vm.constants);
        gc->sweepObjects();
    }
}

File: python/VMPrototype1/repl.py
class Repl:
    def __init__(self, parser=None, program=None, evaluator=None, compiler=None, vm=None):
        self.parser = parser
        self.program = program
        self.evaluator = evaluator
        self.compiler = compiler
        self.vm = vm

    def execute_file(self, source):
        self.parser.load(source)
        self.program.reset()
        self.parser.build_program_statements(self.program)
        print(str(self.program))

        self.compiler.reset()
        self.compiler.compile(self.program)
        bytecode = self.compiler.get_bytecode()
        self.compiler.code.instructions = bytecode.instructions
        self.compiler.code.instructions.size = len(bytecode.instructions.bytes)
        print(bytecode.instructions.bytes)
        print(self.compiler.code.to_string(), end='')

        self.vm.load(bytecode)
        self.vm.run()
        result = self.vm.last_popped_stack_elem()
        #print(self.vm.stack)
        #print(self.vm.sp)
        #result = self.vm.last_popped_stack_element()
        #if result:
        print(result.inspect())

    def loop(self):
        while True:
            line = input(">> ")
            if line == "exit()":
                break
            if line == "stack()":
                self.vm.print_stack()
                continue
            self.parser.load(line)
            self.program.reset()
            self.parser.build_program_statements(self.program)
            print(str(self.program))

            self.compiler.reset()
            self.compiler.compile(self.program)
            bytecode = self.compiler.get_bytecode()
            self.compiler.code.instructions = bytecode.instructions
            print(self.compiler.code.to_string(), end='')

            self.vm.load(bytecode)
            self.vm.run()
            result = self.vm.last_popped_stack_elem()
            #result = self.vm.last_popped_stack_element()
            #if result:
            print(result.inspect())
